File: desktop/src/features/workstream-board/lib/boardReferences.ts
export const BOARD_REFERENCE_TAG = "buzz:board-ref";
export const BOARD_REFERENCE_VERSION = "1";
export const MAX_BOARD_REFERENCES = 32;

export type BoardPlacement = {
  workstreamId: string;
  workstreamLabel: string;
  sectionId: string;
  sectionLabel: string;
};
export type BoardSnapshot = { label: string; detail?: string };
type Common = {
  identity: string;
  snapshot: BoardSnapshot;
  placement: BoardPlacement;
};
export type BoardReference =
  | (Common & { kind: "workstream" })
  | (Common & { kind: "pull-request" })
  | (Common & { kind: "agent" })
  | (Common & { kind: "agent-group" })
  | (Common & {
      kind: "thread-blocker";
      destination: {
        channelId: string;
        messageId: string;
        threadRootId?: string;
      };
    });

const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
const HEX = /^[0-9a-f]{64}$/;
const KINDS = new Set([
  "workstream",
  "pull-request",
  "agent",
  "agent-group",
  "thread-blocker",
]);
const safe = (value: unknown, max = 160): value is string =>
  typeof value === "string" &&
  value.trim().length > 0 &&
  value.length <= max &&
  ![...value].some((character) => {
    const code = character.charCodeAt(0);
    return code < 32 || code === 127;
  });

export function isBoardReference(value: unknown): value is BoardReference {
  if (!value || typeof value !== "object" || Array.isArray(value)) return false;
  const v = value as Record<string, unknown>;
  if (
    typeof v.kind !== "string" ||
    !KINDS.has(v.kind) ||
    !safe(v.identity, 512) ||
    !v.snapshot ||
    typeof v.snapshot !== "object" ||
    Array.isArray(v.snapshot) ||
    !safe((v.snapshot as Record<string, unknown>).label) ||
    ((v.snapshot as Record<string, unknown>).detail !== undefined &&
      !safe((v.snapshot as Record<string, unknown>).detail))
  )
    return false;
  const p = v.placement as Record<string, unknown> | undefined;
  if (
    !p ||
    !safe(p.workstreamId, 36) ||
    !UUID.test(p.workstreamId) ||
    !safe(p.workstreamLabel) ||
    !safe(p.sectionId, 512) ||
    !safe(p.sectionLabel)
  )
    return false;
  if (v.kind === "agent" && !HEX.test(v.identity)) return false;
  if (v.kind === "thread-blocker") {
    const d = v.destination as Record<string, unknown> | undefined;
    if (
      !d ||
      !safe(d.channelId, 36) ||
      !UUID.test(d.channelId) ||
      !safe(d.messageId, 64) ||
      !HEX.test(d.messageId) ||
      (d.threadRootId !== undefined &&
        (!safe(d.threadRootId, 64) || !HEX.test(d.threadRootId)))
    )
      return false;
  }
  return true;
}

export function boardReferenceTags(
  references: readonly BoardReference[],
): string[][] {
  if (references.length > MAX_BOARD_REFERENCES)
    throw new Error(`Select at most ${MAX_BOARD_REFERENCES} references.`);
  const seen = new Set<string>();
  return references.map((reference) => {
    if (!isBoardReference(reference) || seen.has(reference.identity))
      throw new Error("Invalid or duplicate Board reference.");
    seen.add(reference.identity);
    const payload = JSON.stringify(reference);
    if (new TextEncoder().encode(payload).length > 4096)
      throw new Error("Board reference is too large.");
    return [BOARD_REFERENCE_TAG, BOARD_REFERENCE_VERSION, payload];
  });
}

export function parseBoardReferences(
  tags: readonly (readonly string[])[] | undefined,
): BoardReference[] {
  const result: BoardReference[] = [];
  const seen = new Set<string>();
  for (const tag of tags ?? []) {
    if (result.length >= MAX_BOARD_REFERENCES) break;
    if (
      tag.length !== 3 ||
      tag[0] !== BOARD_REFERENCE_TAG ||
      tag[1] !== BOARD_REFERENCE_VERSION ||
      new TextEncoder().encode(tag[2]).length > 4096
    )
      continue;
    try {
      const value: unknown = JSON.parse(tag[2]);
      if (isBoardReference(value) && !seen.has(value.identity)) {
        seen.add(value.identity);
        result.push(value);
      }
    } catch {
      /* malformed references do not break the message */
    }
  }
  return result;
}

export type BoardReferenceResolution = {
  reference: BoardReference;
  state: "live" | "changed" | "historical";
};
export function resolveBoardReferences(
  references: readonly BoardReference[],
  live: ReadonlyMap<string, BoardReference>,
): BoardReferenceResolution[] {
  return references.map((reference) => {
    const current = live.get(reference.identity);
    return {
      reference,
      state: !current
        ? "historical"
        : JSON.stringify(current.snapshot) ===
              JSON.stringify(reference.snapshot) &&
            JSON.stringify(current.placement) ===
              JSON.stringify(reference.placement)
          ? "live"
          : "changed",
    };
  });
}

export function boardReferenceKey(
  reference: Pick<BoardReference, "kind" | "identity">,
): string {
  return `${reference.kind}:${reference.identity}`;
}

File: crates/buzz-sdk/src/board_references.rs
//! Typed, bounded references attached to Board discussion messages.

use nostr::Tag;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

/// Event tag name carrying a Board reference.
pub const BOARD_REFERENCE_TAG: &str = "buzz:board-ref";
/// Current wire version.
pub const BOARD_REFERENCE_VERSION: &str = "1";
/// Maximum references accepted on one message.
pub const MAX_BOARD_REFERENCES: usize = 32;
const MAX_PAYLOAD_BYTES: usize = 4096;
const MAX_LABEL_BYTES: usize = 160;
const MAX_ID_BYTES: usize = 512;

/// As-sent placement used when a live object no longer exists.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct BoardPlacement {
    /// Canonical workstream channel UUID.
    pub workstream_id: String,
    /// Least-privilege workstream display label.
    pub workstream_label: String,
    /// Stable Board section identifier.
    pub section_id: String,
    /// As-sent section display label.
    pub section_label: String,
}

/// Immutable semantic snapshot safe to show after live data disappears.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct BoardSnapshot {
    /// Primary display label.
    pub label: String,
    /// Optional secondary state/relationship annotation.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
}

/// Canonical thread destination bound to a blocker.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct BoardThreadDestination {
    /// Destination channel UUID.
    pub channel_id: String,
    /// Exact target message ID.
    pub message_id: String,
    /// Thread root ID when the target is in a thread.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub thread_root_id: Option<String>,
}

/// Supported Board domain terms. `kind` is the discriminant on the wire.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "kebab-case", deny_unknown_fields)]
pub enum BoardReference {
    /// A workstream card.
    Workstream {
        /// Canonical channel UUID.
        identity: String,
        /// As-sent display data.
        snapshot: BoardSnapshot,
        /// As-sent Board placement.
        placement: BoardPlacement,
    },
    /// A Buzz or GitHub pull request.
    PullRequest {
        /// Canonical provider-qualified identity.
        identity: String,
        /// As-sent display data.
        snapshot: BoardSnapshot,
        /// As-sent Board placement.
        placement: BoardPlacement,
    },
    /// An agent identity.
    Agent {
        /// Canonical lowercase hex pubkey.
        identity: String,
        /// As-sent display data.
        snapshot: BoardSnapshot,
        /// As-sent Board placement.
        placement: BoardPlacement,
    },
    /// A stable agent-team identity.
    AgentGroup {
        /// Canonical team identity.
        identity: String,
        /// As-sent display data.
        snapshot: BoardSnapshot,
        /// As-sent Board placement.
        placement: BoardPlacement,
    },
    /// A blocker bound to an exact conversation destination.
    ThreadBlocker {
        /// Stable manual wait key or canonical PR-derived wait identity.
        identity: String,
        /// Exact conversation destination.
        destination: BoardThreadDestination,
        /// As-sent display data.
        snapshot: BoardSnapshot,
        /// As-sent Board placement.
        placement: BoardPlacement,
    },
}

impl BoardReference {
    fn dedup_key(&self) -> (&'static str, String) {
        match self {
            Self::Workstream { identity, .. } => ("workstream", identity.clone()),
            Self::PullRequest { identity, .. } => ("pull-request", identity.clone()),
            Self::Agent { identity, .. } => ("agent", identity.clone()),
            Self::AgentGroup { identity, .. } => ("agent-group", identity.clone()),
            Self::ThreadBlocker { identity, .. } => ("thread-blocker", identity.clone()),
        }
    }

    fn identity(&self) -> &str {
        match self {
            Self::Workstream { identity, .. }
            | Self::PullRequest { identity, .. }
            | Self::Agent { identity, .. }
            | Self::AgentGroup { identity, .. }
            | Self::ThreadBlocker { identity, .. } => identity,
        }
    }

    fn snapshot(&self) -> &BoardSnapshot {
        match self {
            Self::Workstream { snapshot, .. }
            | Self::PullRequest { snapshot, .. }
            | Self::Agent { snapshot, .. }
            | Self::AgentGroup { snapshot, .. }
            | Self::ThreadBlocker { snapshot, .. } => snapshot,
        }
    }

    fn placement(&self) -> &BoardPlacement {
        match self {
            Self::Workstream { placement, .. }
            | Self::PullRequest { placement, .. }
            | Self::Agent { placement, .. }
            | Self::AgentGroup { placement, .. }
            | Self::ThreadBlocker { placement, .. } => placement,
        }
    }

    fn validate(&self) -> bool {
        let placement = self.placement();
        let snapshot = self.snapshot();
        valid_id(self.identity())
            && valid_uuid(&placement.workstream_id)
            && valid_label(&placement.workstream_label)
            && valid_id(&placement.section_id)
            && valid_label(&placement.section_label)
            && valid_label(&snapshot.label)
            && snapshot.detail.as_deref().is_none_or(valid_label)
            && match self {
                Self::Agent { identity, .. } => valid_hex64(identity),
                Self::ThreadBlocker { destination, .. } => {
                    valid_uuid(&destination.channel_id)
                        && valid_hex64(&destination.message_id)
                        && destination
                            .thread_root_id
                            .as_deref()
                            .is_none_or(valid_hex64)
                }
                _ => true,
            }
    }
}

fn valid_label(value: &str) -> bool {
    !value.trim().is_empty()
        && value.len() <= MAX_LABEL_BYTES
        && !value.chars().any(char::is_control)
}

fn valid_id(value: &str) -> bool {
    !value.trim().is_empty() && value.len() <= MAX_ID_BYTES && !value.chars().any(char::is_control)
}

fn valid_hex64(value: &str) -> bool {
    value.len() == 64
        && value
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

fn valid_uuid(value: &str) -> bool {
    uuid::Uuid::parse_str(value).is_ok() && value == value.to_ascii_lowercase()
}

/// Build validated ordered event tags. Invalid input rejects the whole send.
pub fn build_board_reference_tags(
    references: &[BoardReference],
) -> Result<Vec<Tag>, crate::SdkError> {
    if references.len() > MAX_BOARD_REFERENCES {
        return Err(crate::SdkError::InvalidInput(format!(
            "too many Board references (max {MAX_BOARD_REFERENCES})"
        )));
    }
    let mut identities = HashSet::new();
    references
        .iter()
        .map(|reference| {
            if !reference.validate() || !identities.insert(reference.dedup_key()) {
                return Err(crate::SdkError::InvalidInput(
                    "invalid or duplicate Board reference".into(),
                ));
            }
            let payload = serde_json::to_string(reference)
                .map_err(|error| crate::SdkError::InvalidInput(error.to_string()))?;
            if payload.len() > MAX_PAYLOAD_BYTES {
                return Err(crate::SdkError::InvalidInput(
                    "Board reference payload too large".into(),
                ));
            }
            Tag::parse([
                BOARD_REFERENCE_TAG,
                BOARD_REFERENCE_VERSION,
                payload.as_str(),
            ])
            .map_err(|error| crate::SdkError::InvalidTag(error.to_string()))
        })
        .collect()
}

/// Parse valid references in persisted order. Bad entries fail closed individually.
pub fn parse_board_references(tags: &nostr::Tags) -> Vec<BoardReference> {
    let mut identities = HashSet::new();
    tags.iter()
        .filter_map(|tag| {
            let values = tag.as_slice();
            if values.len() != 3
                || values[0] != BOARD_REFERENCE_TAG
                || values[1] != BOARD_REFERENCE_VERSION
                || values[2].len() > MAX_PAYLOAD_BYTES
            {
                return None;
            }
            let reference: BoardReference = serde_json::from_str(&values[2]).ok()?;
            (reference.validate() && identities.insert(reference.dedup_key())).then_some(reference)
        })
        .take(MAX_BOARD_REFERENCES)
        .collect()
}

/// Parse references authorized for one current Board workstream.
pub fn parse_board_references_for_workstream(
    tags: &nostr::Tags,
    workstream_id: &str,
) -> Vec<BoardReference> {
    parse_board_references(tags)
        .into_iter()
        .filter(|reference| reference.placement().workstream_id == workstream_id)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use nostr::Tags;

    fn reference(identity: &str) -> BoardReference {
        BoardReference::Workstream {
            identity: identity.into(),
            snapshot: BoardSnapshot {
                label: "Checkout recovery".into(),
                detail: Some("Active".into()),
            },
            placement: BoardPlacement {
                workstream_id: "123e4567-e89b-12d3-a456-426614174000".into(),
                workstream_label: "checkout-recovery".into(),
                section_id: "workstream".into(),
                section_label: "Workstream".into(),
            },
        }
    }

    #[test]
    fn round_trip_preserves_order() {
        let tags = build_board_reference_tags(&[reference("one"), reference("two")]).unwrap();
        let parsed = parse_board_references(&Tags::from_list(tags));
        assert_eq!(parsed, vec![reference("one"), reference("two")]);
    }

    #[test]
    fn malformed_unknown_and_duplicate_entries_fail_closed() {
        let mut tags = build_board_reference_tags(&[reference("one")]).unwrap();
        tags.push(Tag::parse([BOARD_REFERENCE_TAG, "99", "{}"]).unwrap());
        tags.push(Tag::parse([BOARD_REFERENCE_TAG, BOARD_REFERENCE_VERSION, "not-json"]).unwrap());
        tags.extend(build_board_reference_tags(&[reference("one"), reference("two")]).unwrap());
        assert_eq!(
            parse_board_references(&Tags::from_list(tags)),
            vec![reference("one"), reference("two")]
        );
    }
    #[test]
    fn cross_workstream_references_fail_closed() {
        let allowed = reference("allowed");
        let mut other = reference("other");
        if let BoardReference::Workstream { placement, .. } = &mut other {
            placement.workstream_id = "123e4567-e89b-12d3-a456-426614174001".into();
        }
        let tags = Tags::from_list(build_board_reference_tags(&[allowed.clone(), other]).unwrap());
        assert_eq!(
            parse_board_references_for_workstream(&tags, "123e4567-e89b-12d3-a456-426614174000"),
            vec![allowed]
        );
    }
}
